hoot-core/src/main/cpp/hoot/core/io/HootNetworkRequest.cpp

```diff
@@ -150,19 +156,48 @@ bool HootNetworkRequest::_networkRequest(const QUrl& url, int timeout,
     return false;
     break;
   }
-  //  Wait for finished signal from reply object
-  _blockOnReply(reply, timeout);
-  //  Get the status and content of the reply if available
-  _status = _getHttpResponseCode(reply);
-  //  According to the documention this shouldn't ever happen
-  if (reply == NULL)
+  //  Start the timer
+  timeoutTimer.start(timeout * 1000);
+  //  Connect the timeout lambda
+  QObject::connect(&timeoutTimer, &QTimer::timeout, [&]()
+  {
+    //  Stop the timer first so the lambda isn't called again
+    timeoutTimer.stop();
+    //  Set the timed out flag before aborting the reply
+    _timedOut = true;
+    //  Abort will trigger QNetworkAccessManager::finished()
+    reply->abort();
+    //  Set the status to negative TimeoutError, negative so that it doesn't
+    //  collide with HTTP status numbers
+    _status = -1 * (int)QNetworkReply::TimeoutError;
+    _error = QString("HTTP Request to %1 timed-out.").arg(tempUrl.toString(QUrl::RemoveUserInfo));
+    //  Exit with a non-zero status
+    loop.exit(_status);
+  });
+  //  Connect the finished lambda
+//  QObject::connect(pNAM.get(), &QNetworkAccessManager::finished, [&]()
+  QObject::connect(reply, &QNetworkReply::finished, [&]()
   {
-    throw HootException(QString("Network request error: GET/POST/PUT request failed to create reply object."));
+    //  The finished signal is emitted when the reply is aborted in the
+    //  timeout lambda, so don't do anything
+    if (!_timedOut)
+    {
+      //  Stop the timer
+      timeoutTimer.stop();
+      _timedOut = false;
+      //  Exit with a zero status
+      loop.exit();
+    }
+  });
+  //  Actually run the event loop waiting for either lambda to exit the loop
+  if (loop.exec() != 0)
     return false;
-  }
+  //  Read all of the data
   _content = reply->readAll();
+  //  Get the status and content of the reply if available
+  _status = _getHttpResponseCode(reply);
   //  Check error status on our reply
-  if (QNetworkReply::NoError != reply->error())
+  if (_status == 0 && QNetworkReply::NoError != reply->error())
```

Comment:
HTTP error codes that are received actually trigger an error in the HTTP handling of the `reply` object so `reply->error()` returns an error code.  The following error handling code is only for errors that `_getHttpResponseCode` cannot parse out of the HTTP response (i.e. timeouts, etc.).